### PR TITLE
[clang] NFCI: use TemplateArgumentLoc for NTTP DefaultArgument

### PR DESCRIPTION
**Pull request by mizvekov**
This is an enabler for https://github.com/llvm/llvm-project/pull/92855

This allows an NTTP default argument to be set as an arbitrary TemplateArgument, not just an expression.
This allows template parameter packs to have default arguments in the AST, even though the language proper doesn't support the syntax for it.

This allows NTTP default arguments to be other kinds of arguments, like packs, integral constants, and such.

### CODE REVIEW

File: clang/include/clang/AST/DeclTemplate.h [hunk 1]
```diff
@@ -1435,7 +1436,10 @@ class NonTypeTemplateParmDecl final
   bool hasDefaultArgument() const { return DefaultArgument.isSet(); }
 
   /// Retrieve the default argument, if any.
-  Expr *getDefaultArgument() const { return DefaultArgument.get(); }
+  const TemplateArgumentLoc &getDefaultArgument() const {
```
Comment (ChuanqiXu9) [inline clang/include/clang/AST/DeclTemplate.h:1439]: Is it better to rename `getDefaultArgument` to `getDefaultArgumentLoc`?

**Review by ChuanqiXu9 — COMMENTED** [review]
Can you try to explain the what the patch does more and describe the rationale?

File: clang/include/clang/AST/DeclTemplate.h (hunk 1 above)
Comment (mizvekov) [reply clang/include/clang/AST/DeclTemplate.h:1439]: All template parameters currently use that name, so this is kept for consistency.

I wouldn't object a later separate patch to rename all of them at once.

**Comment by mizvekov** [comment]
> Can you try to explain the what the patch does more and describe the rationale?

Done.

**Comment by mizvekov** [comment]
@erichkeane this is kind of the same as https://github.com/llvm/llvm-project/pull/92854, but for NTTP instead of type-parameters.

**Review by erichkeane — APPROVED** [review]

**Pull request merged**